libs/common/src/test-helpers/schemas/with-upsert-tfm-user-request-schema.tests.ts

```diff
@@ -0,0 +1,39 @@
+import { ZodSchema } from 'zod';
+import { withSchemaTests } from './with-schema.tests';
+import { getCreateTfmUserRequestFailureTestCases, getCreateTfmUserRequestSuccessTestCases } from './with-create-tfm-user-request-schema.tests';
+
+/**
+ * This is a reusable test to allow for complete testing of schemas that
+ * utilise the UPSERT_TFM_USER_REQUEST_SCHEMA as part of their definition
+ *
+ * Note: UPSERT_TFM_USER_REQUEST_SCHEMA is effectively an alias for CREATE_TFM_USER_REQUEST_SCHEMA
+ */
+
+type TestCasesParams = {
+  getTestObjectWithUpdatedUpsertTfmUserRequestParams: (upsertTfmUserRequest: unknown) => unknown;
+};
+
+type WithUpsertTfmUserRequestSchemaTestsParams = {
+  schema: ZodSchema;
+} & Partial<TestCasesParams>;
+
+export function withUpsertTfmUserRequestSchemaTests({
+  schema,
+  getTestObjectWithUpdatedUpsertTfmUserRequestParams = (upsertTfmUserRequest) => upsertTfmUserRequest,
+}: WithUpsertTfmUserRequestSchemaTestsParams) {
+  describe('with UPSERT_TFM_USER_REQUEST_SCHEMA tests', () => {
+    withSchemaTests({
+      schema,
+      failureTestCases: getUpsertTfmUserRequestFailureTestCases({ getTestObjectWithUpdatedUpsertTfmUserRequestParams }),
+      successTestCases: getUpsertTfmUserRequestSuccessTestCases({ getTestObjectWithUpdatedUpsertTfmUserRequestParams }),
+    });
+  });
+}
+
+export function getUpsertTfmUserRequestFailureTestCases({ getTestObjectWithUpdatedUpsertTfmUserRequestParams }: TestCasesParams) {
```

Comment:
(This does have the additional cognetive load of piggybacking off of `create` but it should do to keep tests DRY)

Reply:
Any suggestions for this? I'm not sure how we could as I think each segment of this provides important context (and probably doesnt want changing -- for instance if we changed the part `UpsertTfmUserRequest`, it'd add abiguity as it would not be identifiable to the `UpsertTfmUserRequest` type without having to open and check this file)?

Reply:
@ttbarnes I've had a thought about this -- I'm not sure I can reduce `getTestObjectWithUpdatedCreateTfmUserRequestParams: getTestObjectWithUpdatedUpsertTfmUserRequestParams` any further either -- with `getTestObjectWithUpdatedCreateTfmUserRequestParams` being because we're need a function that gets a test object with updated `CreateTfmUserRequest`...And we need that to be a function to pass in how we set whatever param is needing to be set by the `updatedUpsertTfmUserRequest` object...